src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -0,0 +1,1305 @@
+        {
+            return this;
+        }
+
+        // The documents might be contributing to multiple different projects; split them by project and then we'll process
+        // project-at-a-time.
+        var documentIdsByProjectId = documentIds.ToLookup(id => id.ProjectId);
+
+        var newCompilationState = this;
+
+        foreach (var documentIdsInProject in documentIdsByProjectId)
+        {
+            var oldProjectState = this.Solution.GetProjectState(documentIdsInProject.Key);
+
+            if (oldProjectState == null)
+            {
+                throw new InvalidOperationException(string.Format(WorkspacesResources._0_is_not_part_of_the_workspace, documentIdsInProject.Key));
+            }
+
+            var removedDocumentStatesBuilder = ArrayBuilder<T>.GetInstance();
+
+            foreach (var documentId in documentIdsInProject)
+            {
+                removedDocumentStatesBuilder.Add(getExistingTextDocumentState(oldProjectState, documentId));
+            }
+
+            var removedDocumentStatesForProject = removedDocumentStatesBuilder.ToImmutableAndFree();
+
+            var (newProjectState, compilationTranslationAction) = removeDocumentsFromProjectState(oldProjectState, documentIdsInProject.ToImmutableArray(), removedDocumentStatesForProject);
+
+            (var newSolutionState, _, _) = newCompilationState.Solution.ForkProject(
+                oldProjectState,
+                newProjectState,
+                // Intentionally using this.Solution here and not newSolutionState
+                newFilePathToDocumentIdsMap: this.Solution.CreateFilePathToDocumentIdsMapWithRemovedDocuments(removedDocumentStatesForProject));
+
+            newCompilationState = newCompilationState.ForkProject(
+                newSolutionState,
+                newProjectState,
+                compilationTranslationAction,
+                forkTracker: true);
+        }
+
+        return newCompilationState;
+    }
+
+    /// <inheritdoc cref="Solution.WithCachedSourceGeneratorState(ProjectId, Project)"/>
+    public SolutionCompilationState WithCachedSourceGeneratorStateWorker(ProjectId projectToUpdate, Project projectWithCachedGeneratorState)
+    {
+        this.Solution.CheckContainsProject(projectToUpdate);
+
+        // First see if we have a generator driver that we can get from the other project.
+
+        if (!projectWithCachedGeneratorState.Solution.CompilationState.TryGetCompilationTracker(projectWithCachedGeneratorState.Id, out var tracker) ||
+            tracker.GeneratorDriver is null)
+        {
+            // We don't actually have any state at all, so no change.
+            return this;
+        }
+
+        var projectToUpdateState = this.Solution.GetRequiredProjectState(projectToUpdate);
+
+        // Note: we have to force this fork to happen as the actual solution-state object is not changing.
+        var newCompilationState = this.ForceForkProject(
+            this.Solution,
```

Comment:
this is an interesting case.  we're just adding state related to a particular project.  So the actual solution-state snapshot isn't changing.  so we have to call through ForceForkProject so we dont' insta-bail.